### ISSUE TITLE
No Sound in different system and platform

### ISSUE THREAD

**Issue by wowotek** [waifu-motivator/waifu-motivator-plugin#259]
**Describe the bug**
In Ubuntu 20.04 Fresh install, some people (including me) experiencing no sound issue.

**To Reproduce**
1. Install Ubuntu 20.04 fresh from download
2. Install Pycharm from snap
    ```shell
    sudo snap install pycharm-community --classic
    ```
3. Install **Waifu Motifator** from `file` > `settings` > `Plugins` > search for Waifu Motivator > install it
4. Restart Pycharm
5. press `Alt` + `M` with intention to motivate your self
6. Voila ! No Sound :(

**Expected behavior**

Waifu Motivator should play the sound correctly and accordingly.

**Screenshots**

There is no picture of a sound

**Platform (please complete the following information):**

 - OS: Linux 5.4.0-52-generic - Ubuntu 20.04 focal
 - IDE Pycharm - Community
 - IDE Version 2020.2.3 0- Build #PC-202.7660.27
 - Version WMP 1.3

**Additional context**

I'm no linux master but i suppose it's because WMP (or pycharm, i dunno) play the sound file with FFmpeg.
turns out my(or all) Ubuntu 20.04 install don't come with FFmpeg in it.
So I'm Successfully Fixed it by
1. installing FFmpeg with:
    ```shell
    sudo apt-get install ffmpeg
    ```
2. And i make sure i have the sound asset.
    ```shell
    find ~/.config/JetBrains/PyCharmCE2020.2/waifuMotivationAssets/audible/
    ```
3. try to play one of the sound
    ```shell
    cd ~/.config/JetBrains/PyCharmCE2020.2/waifuMotivationAssets/audible/celebration/
    ffplay -nodisp -autoexit good_job.wav
    ```
4. if the sound played successfully, restart Pycharm
5. `Alt` + `M` to motivate yourself

**Comment by Unthrottled** [comment]
> I'm no linux master but i suppose it's because WMP (or pycharm, i dunno) play the sound file with FFmpeg.

I'm running Ubuntu 20.04 as well (that's my main dev environment)! 
I've removed `ffmpeg` and the sound still plays in Pycharm. So maybe I'll need to try spinning up a virtual machine to try to reproduce the issue.

**Comment by zd-zero** [comment]
I tried to reproduce it in a VM and I can consistently reproduce it by following the steps, here's the stack trace that I'm getting.

<details>
<summary>Stacktrace</summary>

```java
java.lang.IllegalArgumentException: No line matching interface Clip supporting format PCM_SIGNED unknown sample rate, 16 bit, stereo, 4 bytes/frame, big-endian is supported.
    at java.desktop/javax.sound.sampled.AudioSystem.getLine(AudioSystem.java:425)
    at java.desktop/javax.sound.sampled.AudioSystem.getClip(AudioSystem.java:461)
    at zd.zero.waifu.motivator.plugin.player.SoundClipUtil.openClip(SoundClipUtil.java:11)
    at zd.zero.waifu.motivator.plugin.player.DefaultWaifuSoundPlayer.play(DefaultWaifuSoundPlayer.java:33)
    at zd.zero.waifu.motivator.plugin.motivation.BaseMotivation.soundAlert(BaseMotivation.kt:83)
    at zd.zero.waifu.motivator.plugin.motivation.BaseMotivation.motivate(BaseMotivation.kt:33)
    at zd.zero.waifu.motivator.plugin.actions.MotivateMeAction.lambda$actionPerformed$0(MotivateMeAction.java:44)
    at zd.zero.waifu.motivator.plugin.tools.ToolBox$doOrElse$1.invoke(ToolBox.kt:28)
    at zd.zero.waifu.motivator.plugin.tools.ToolBox$doOrElse$1.invoke(ToolBox.kt:19)
    at zd.zero.waifu.motivator.plugin.tools.ExtensionsKt$doOrElse$4.accept(Extensions.kt:30)
    at zd.zero.waifu.motivator.plugin.tools.ExtensionsKt$doOrElse$4.accept(Extensions.kt)
    at java.base/java.util.Optional.ifPresent(Optional.java:183)
    at zd.zero.waifu.motivator.plugin.tools.ExtensionsKt.doOrElse(Extensions.kt:28)
    at zd.zero.waifu.motivator.plugin.tools.ToolBox.doOrElse(ToolBox.kt:28)
    at zd.zero.waifu.motivator.plugin.actions.MotivateMeAction.lambda$actionPerformed$2(MotivateMeAction.java:33)
    at com.intellij.util.RunnableCallable.call(RunnableCallable.java:20)
    at com.intellij.util.RunnableCallable.call(RunnableCallable.java:11)
    at com.intellij.openapi.application.impl.ApplicationImpl$1.call(ApplicationImpl.java:268)
    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
    at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:668)
    at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:665)
    at java.base/java.security.AccessController.doPrivileged(Native Method)
    at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:665)
    at java.base/java.lang.Thread.run(Thread.java:834)
```
</details>

**Comment by wowotek** [comment]
_**TLDR**; currently it's working with or without `FFmpeg`._

With ffmpeg installed, stacktrace after welcoming gif and sound
```
2020-10-30 18:53:39,404 [   8566]   WARN - Container.ComponentManagerImpl - Do not use constructor injection (requestorClass=mobi.hsz.idea.gitignore.IgnoreFileBasedIndexProjectHandler) 
2020-10-30 18:53:41,194 [  10356]   WARN - tartup.impl.StartupManagerImpl - Activities registered via registerPostStartupActivity must be dumb-aware: zd.zero.waifu.motivator.plugin.WaifuMotivatorProject$$Lambda$1069/0x0000000100bd2840@411a6c9d 
2020-10-30 18:53:46,229 [  15391]   WARN - Container.ComponentManagerImpl - Do not use constructor injection (requestorClass=mobi.hsz.idea.gitignore.daemon.AddUnversionedFilesNotificationProvider) 
2020-10-30 18:53:46,233 [  15395]   WARN - Container.ComponentManagerImpl - Do not use constructor injection (requestorClass=mobi.hsz.idea.gitignore.daemon.MissingGitignoreNotificationProvider) 
2020-10-30 18:53:57,435 [  26597]   WARN - .skeletons.PySkeletonRefresher - Some skeletons failed to generate 
2020-10-30 18:53:57,436 [  26598]   WARN - .skeletons.PySkeletonRefresher - pydantic.__init__ 
2020-10-30 18:53:57,436 [  26598]   WARN - .skeletons.PySkeletonRefresher - pydantic.mypy
```

After that i uninstalled ffmpeg via
```
sudo apt-get remove ffmpeg
```
and have confirmation that i cannot run `ffmpeg` and `ffplay`

and then i ran the pycharm and here's the stacktrace after the welcoming gif and sound.
```
2020-10-30 18:48:20,147 [  10072]   WARN - Container.ComponentManagerImpl - Do not use constructor injection (requestorClass=mobi.hsz.idea.gitignore.IgnoreFileBasedIndexProjectHandler) 
2020-10-30 18:48:21,866 [  11791]   WARN - tartup.impl.StartupManagerImpl - Activities registered via registerPostStartupActivity must be dumb-aware: zd.zero.waifu.motivator.plugin.WaifuMotivatorProject$$Lambda$1082/0x0000000100c1d840@5aa4017b 
2020-10-30 18:48:27,616 [  17541]   WARN - Container.ComponentManagerImpl - Do not use constructor injection (requestorClass=mobi.hsz.idea.gitignore.daemon.AddUnversionedFilesNotificationProvider) 
2020-10-30 18:48:27,627 [  17552]   WARN - Container.ComponentManagerImpl - Do not use constructor injection (requestorClass=mobi.hsz.idea.gitignore.daemon.MissingGitignoreNotificationProvider) 
2020-10-30 18:48:42,754 [  32679]   WARN - .skeletons.PySkeletonRefresher - Some skeletons failed to generate 
2020-10-30 18:48:42,755 [  32680]   WARN - .skeletons.PySkeletonRefresher - pydantic.__init__ 
2020-10-30 18:48:42,755 [  32680]   WARN - .skeletons.PySkeletonRefresher - pydantic.mypy
```

i cannot invest any time right now to replicate again. but I have my new PC parts upgrade in 3 days. since i do really don't have any sound before. i will try again with really fresh install Ubuntu 20.04. and here's the runs-down of what i'm gonna do.

- `/dev/null` all my drive
- Install Ubuntu 20.04 (The _**NOT**_ minimal installation)
- Check if i can run `FFmpeg` or `FFplay`
- Install Pycharm immediately
- Open Pycharm via terminal
- Install WMP immediately via plugin browser (the `settings` under `File` menu)
- Copy, and save the stacktrace
- Check the sound asset under `~/.config`
- Alt + M Feature, copy and save the stacktrace
- Restart Pycharm
- Wait for Greetings sound
- Copy, and save the stacktrace
- Close Pycharm
- Install FFmpeg
- Open Pycharm Via terminal, and wait for the Greetings
- Copy, and save the stacktrace
- Alt + M Feature, copy and save the stacktrace
- Wait for Greetings sound
- Copy, and save the stacktrace

is that good. or you have anything that i should do ?

**Comment by zd-zero** [comment]
Yep, I think that would be fine. I guess we'll leave this open for now and once you've confirmed that it's working, we'll close this. Thank you for spending some time with this, so we can ensure that all devs can hear a good Sayonara ;)

**Comment by Unthrottled** [comment]
:thinking: 

@zd-zero I think I saw the issue you reported (It included the issue number in the additional info)

https://sentry.io/share/issue/e17168135a844d7e95d8ab9171389ed3/

However, it was also reported by another user on windows

**Additional data**

Additional Info | None
-- | --
App Name | PyCharm 2020.2.3 (Community Edition)
Build Info | Build #PC-202.7660.27, built on October 6, 2020
Cores | 2
Current LAF | Darcula
GC | ParNew,ConcurrentMarkSweep
JRE | Runtime version: 11.0.8+10-b944.34 amd64
Memory | 733
Message | Error executing task zd.zero.waifu.motivator.plugin.WaifuMotivatorProject$$Lambda$1617/0x00000001011e7c40@9848d1
Non-Bundled Plugins | IdeaVIM,codes.rudolph.idea.cfg,com.4lex4.intellij.solarized,com.chrisrm.idea.MaterialThemeUI,com.intellij.ideolog,com.intellij.marketplace,com.jetbrains.CyanTheme,detekt,net.seesharpsoft.intellij.plugins.csv,Dart,some.awesome,zd.zero.waifu-motivator-plugin,aws.toolkit,com.illuminatedcloud2.intellij,ru.adelf.idea.dotenv,R4Intellij,tech.lnkd.LNKD-tech-Editor
Plugin Config | {   "isWaifuOfTheDayEnabled": false,   "isStartupMotivationEnabled": true,   "isStartupMotivationSoundEnabled": true,   "isUnitTesterMotivationEnabled": true,   "isUnitTesterMotivationSoundEnabled": true,   "isMotivateMeEnabled": true,   "isMotivateMeSoundEnabled": true,   "isSayonaraEnabled": true,   "isDisabledInDistractionFreeMode": true,   "isTaskMotivationEnabled": true,   "isTaskSoundEnab...
Registry | ide.balloon.shadow.size=0
System Info | Windows 7 6.1
VM | VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.

**Comment by zd-zero** [comment]
> However, it was also reported by another user on windows

If that's the case, maybe we could start looking into a different library for playing sounds 🕵️‍♂️. I'm thinking of not replacing the current players, we can implement a new player and queue them up, if one player can't play something, they'll pass it to the next player, so we exhaust our players before giving up 😂

**Comment by Unthrottled** [comment]
> Pass to the next player
![Hot potato](https://media.giphy.com/media/Jog9hFsni6hKqm6z9z/giphy.gif)

**Comment by wowotek** [comment]
Hello, after doing the the run down exactly. the output is exactly the same like before.

i don't find any issue anymore. i don't know what's happening before. but thanks for y'all time

> _**TLDR**; currently it's working with or without `FFmpeg`._
> 
> With ffmpeg installed, stacktrace after welcoming gif and sound
> 
> ```
> 2020-10-30 18:53:39,404 [   8566]   WARN - Container.ComponentManagerImpl - Do not use constructor injection (requestorClass=mobi.hsz.idea.gitignore.IgnoreFileBasedIndexProjectHandler) 
> 2020-10-30 18:53:41,194 [  10356]   WARN - tartup.impl.StartupManagerImpl - Activities registered via registerPostStartupActivity must be dumb-aware: zd.zero.waifu.motivator.plugin.WaifuMotivatorProject$$Lambda$1069/0x0000000100bd2840@411a6c9d 
> 2020-10-30 18:53:46,229 [  15391]   WARN - Container.ComponentManagerImpl - Do not use constructor injection (requestorClass=mobi.hsz.idea.gitignore.daemon.AddUnversionedFilesNotificationProvider) 
> 2020-10-30 18:53:46,233 [  15395]   WARN - Container.ComponentManagerImpl - Do not use constructor injection (requestorClass=mobi.hsz.idea.gitignore.daemon.MissingGitignoreNotificationProvider) 
> 2020-10-30 18:53:57,435 [  26597]   WARN - .skeletons.PySkeletonRefresher - Some skeletons failed to generate 
> 2020-10-30 18:53:57,436 [  26598]   WARN - .skeletons.PySkeletonRefresher - pydantic.__init__ 
> 2020-10-30 18:53:57,436 [  26598]   WARN - .skeletons.PySkeletonRefresher - pydantic.mypy
> ```
> 
> After that i uninstalled ffmpeg via
> 
> ```
> sudo apt-get remove ffmpeg
> ```
> 
> and have confirmation that i cannot run `ffmpeg` and `ffplay`
> 
> and then i ran the pycharm and here's the stacktrace after the welcoming gif and sound.
> 
> ```
> 2020-10-30 18:48:20,147 [  10072]   WARN - Container.ComponentManagerImpl - Do not use constructor injection (requestorClass=mobi.hsz.idea.gitignore.IgnoreFileBasedIndexProjectHandler) 
> 2020-10-30 18:48:21,866 [  11791]   WARN - tartup.impl.StartupManagerImpl - Activities registered via registerPostStartupActivity must be dumb-aware: zd.zero.waifu.motivator.plugin.WaifuMotivatorProject$$Lambda$1082/0x0000000100c1d840@5aa4017b 
> 2020-10-30 18:48:27,616 [  17541]   WARN - Container.ComponentManagerImpl - Do not use constructor injection (requestorClass=mobi.hsz.idea.gitignore.daemon.AddUnversionedFilesNotificationProvider) 
> 2020-10-30 18:48:27,627 [  17552]   WARN - Container.ComponentManagerImpl - Do not use constructor injection (requestorClass=mobi.hsz.idea.gitignore.daemon.MissingGitignoreNotificationProvider) 
> 2020-10-30 18:48:42,754 [  32679]   WARN - .skeletons.PySkeletonRefresher - Some skeletons failed to generate 
> 2020-10-30 18:48:42,755 [  32680]   WARN - .skeletons.PySkeletonRefresher - pydantic.__init__ 
> 2020-10-30 18:48:42,755 [  32680]   WARN - .skeletons.PySkeletonRefresher - pydantic.mypy
> ```
> 
> i cannot invest any time right now to replicate again. but I have my new PC parts upgrade in 3 days. since i do really don't have any sound before. i will try again with really fresh install Ubuntu 20.04. and here's the runs-down of what i'm gonna do.
> 
>     * `/dev/null` all my drive
> 
>     * Install Ubuntu 20.04 (The _**NOT**_ minimal installation)
> 
>     * Check if i can run `FFmpeg` or `FFplay`
> 
>     * Install Pycharm immediately
> 
>     * Open Pycharm via terminal
> 
>     * Install WMP immediately via plugin browser (the `settings` under `File` menu)
> 
>     * Copy, and save the stacktrace
> 
>     * Check the sound asset under `~/.config`
> 
>     * Alt + M Feature, copy and save the stacktrace
> 
>     * Restart Pycharm
> 
>     * Wait for Greetings sound
> 
>     * Copy, and save the stacktrace
> 
>     * Close Pycharm
> 
>     * Install FFmpeg
> 
>     * Open Pycharm Via terminal, and wait for the Greetings
> 
>     * Copy, and save the stacktrace
> 
>     * Alt + M Feature, copy and save the stacktrace
> 
>     * Wait for Greetings sound
> 
>     * Copy, and save the stacktrace
> 
> 
> is that good. or you have anything that i should do ?

**Comment by zd-zero** [comment]
Awesome! Thank you for your time @wowotek too. We'll leave the issue for now and I'll update it to address the issues encountered by other systems, there are 2 ways we can take, either make all of the sounds into one file format or add one more sound player to handle the sound whenever the other sound-player can't play it. I'm leaning into the former solution so when we allowed users to use their own package, WMP is not limited to one sound format.

**Comment by Unthrottled** [comment]
> I'm leaning into the former solution so when we allowed users to use their own package,  WMP is not limited to one sound format.

That sounds like the latter solution rather than the former solution. Unless I missed something. (Former == 1<sup>st</sup>, Latter == 2<sup>nd</sup>, at least that's how I think it works)

**Comment by zd-zero** [comment]
Right! It's latter day*m, my English is not so good 🙈
